File: Tree Advanced/Tree.cpp
// Practices 6&7. Binary Search Tree Operations
#include <cstdlib>
#include <iostream>
#include <fstream>
#include <sstream>
#include <vector>
#include <tuple>
#include <utility>
#include <algorithm>
#include <queue>
#include <cmath>
using namespace std;
const char BUILD = 'B';
const char FIND_MIN = 'm';
const char FIND_MAX = 'M';
const char SEARCH = 'S';
const char INSERT = 'I';
const char DELETE = 'D';
const char INORDER = 'N';
const char PREORDER = 'R';
const char POSTORDER = 'O';

class TreeNode {
public:
  int key;
  TreeNode* left;
  TreeNode* right;
  TreeNode(int data) {
    key = data;
    left = nullptr;
    right = nullptr;
  }
};

class BinarySearchTree {
public:
  TreeNode* root;
  BinarySearchTree(): root(nullptr) {}
  ~BinarySearchTree();

  TreeNode* arrayToBST(vector<int>&);
  // Return true if tree is empty; false otherwise
  // O(1)
  bool isEmpty() { return (root == nullptr); }
  TreeNode* findMin();
  TreeNode* findMax();
  TreeNode* search(TreeNode*, int);
  void insertNode(int);
  TreeNode* deleteNode(TreeNode*, int);
  void writeInorder(ofstream&, TreeNode*);
  void writePreorder(ofstream&, TreeNode*);
  void writePostorder(ofstream&, TreeNode*);
  void printTree();
private:
  int _getHeight(TreeNode*);
  void _printSpaces(double, TreeNode*);
};

BinarySearchTree::~BinarySearchTree() {
  // Practice 4
  vector<TreeNode*> stack;
  vector<TreeNode*> nodes;
  TreeNode* curr = root;
  TreeNode* prev = nullptr;
  while (curr != nullptr || !stack.empty()) {
    while (curr != nullptr) {
      stack.push_back(curr);
      curr = curr->left;
    }
    curr = stack.back();
    if (curr->right == nullptr || curr->right == prev) {
      stack.pop_back();
      nodes.push_back(curr);
      prev = curr;
      curr = nullptr;
    }
    else {
      curr = curr->right;
    }
  }
  vector<TreeNode*>::iterator it;
  for (it = nodes.begin(); it != nodes.end(); ++it)
    delete *it;
}

// Given a sequence arr of integers, start index l, the end index r, 
// build a binary search (sub)tree that contains keys in arr[l], ..., arr[r].
// Return the root node of the tree
// It is almost similar to divide part in merge sort.
// So time complexity is O(NlogN)
TreeNode* BinarySearchTree::arrayToBST(vector<int>& arr) {
  if (arr.size() == 0) return nullptr;
  for (int i=0; i<arr.size()-1; i++)
    if (arr[i] > arr[i+1]) return nullptr;
  
  int middle = ceil((float)arr.size() / 2) - 1;
  TreeNode* root = new TreeNode(arr[middle]);

  if (arr.size() == 2)
    root->right = new TreeNode(arr[1]);
  else {
    vector<int> left_item(arr.begin(), arr.begin() + middle);
    vector<int> right_item(arr.begin() + middle + 1, arr.end());
    root->left = BinarySearchTree::arrayToBST(left_item);
    root->right = BinarySearchTree::arrayToBST(right_item);
  }

  return root;
}

// Return the node with the minimum value 
// in balanced tree, height is O(logN).
// so time complexity is O(logN)
// Here, N is the number of nodes in tree.
TreeNode* BinarySearchTree::findMin() {
  TreeNode* ptr = root;
  while (ptr->left != nullptr)
    ptr = ptr->left;
  return ptr;
}

// Return the node with the maximum value 
// in balanced tree, height is O(logN).
// so time complexity is O(logN)
// Here, N is the number of nodes in tree.
TreeNode* BinarySearchTree::findMax() {
  TreeNode* ptr = root;
  while (ptr->right != nullptr)
    ptr = ptr->right;
  return ptr;
}

// Given a query, search for the node whose key is equal to query.
// If the node exists, return the key
// Otherwise, return nullptr
// on average, time complexity is O(logN)
// Here, N is the number of nodes in tree.
TreeNode* BinarySearchTree::search(TreeNode* ptr, int query) {
  if (ptr == nullptr) return nullptr;
  // Practice 6
  if (ptr->key == query)
    return ptr;
  else if (ptr->key > query && ptr->left != nullptr)
    return BinarySearchTree::search(ptr->left, query);
  else if (ptr->key < query && ptr->right != nullptr)
    return BinarySearchTree::search(ptr->right, query);
  return ptr;
}

// If node with key k alreay exists in the tree, do nothing
// Otherwise, insert a new node with key k 
// Find location to insert : same with search, so O(logN)
// Adding Nodes : O(1)
// So, time complexity is O(logN)
void BinarySearchTree::insertNode(int k) {
  // TODO. Practice 7
  if (root == nullptr) 
    root = new TreeNode(k);
  else {
    TreeNode* ptr = BinarySearchTree::search(root, k);
    if (k < ptr->key)
      ptr->left = new TreeNode(k);
    else if (k > ptr->key)
      ptr->right = new TreeNode(k);
  }
}
// If deletion fails, immediately terminate the program
// Otherwise, delete the node with key k
// It takes tree's root and key to remove, and returns root node of that subtree.
// When calling, we just need to call like this:
// tree->root = tree->deleteNode(tree->root, key)
// On average, time complexity is O(logN).
// Here, N is the number of nodes in tree.
TreeNode* BinarySearchTree::deleteNode(TreeNode* root_, int k) {
  if (root_ == nullptr) return nullptr;
  if (k < root_->key) 
    root_->left = BinarySearchTree::deleteNode(root_->left, k);
  else if (k > root_->key)
    root_->right = BinarySearchTree::deleteNode(root_->right, k);
  else {
    if (root_->left == nullptr)
      return root_->right;
    else if (root_->right == nullptr)
      return root_->left;
    
    TreeNode* ptr = root_;
    while (ptr->left != nullptr)
      ptr = ptr->left;
    root_->key = ptr->key;
    root_->right = BinarySearchTree::deleteNode(root_->right, root_->key);
  }
  return root_;
}

// Given an output file stream, write the keys of all the nodes 
// visited in inorder traversal
// We need to visit all N nodes, so time complexity is O(N)
void BinarySearchTree::writeInorder(ofstream& outFile, TreeNode* ptr) {
  // Practice 6
  if (ptr != nullptr) {
    BinarySearchTree::writeInorder(outFile, ptr->left);
    outFile << to_string(ptr->key) << " ";
    BinarySearchTree::writeInorder(outFile, ptr->right);
  }
}

// Given an output file stream, write the keys of all the nodes 
// visited in preorder traversal
// We need to visit all N nodes, so time complexity is O(N)
void BinarySearchTree::writePreorder(ofstream& outFile, TreeNode* ptr) {
  // Practice 6
  if (ptr != nullptr) {
    outFile << to_string(ptr->key) << " ";
    BinarySearchTree::writePreorder(outFile, ptr->left);
    BinarySearchTree::writePreorder(outFile, ptr->right);
  }
}

// Given an output file stream, write the keys of all the nodes 
// visited in postorder traversal
// We need to visit all N nodes, so time complexity is O(N)
void BinarySearchTree::writePostorder(ofstream& outFile, TreeNode* ptr) {
  // Practice 6
  if (ptr != nullptr) {
    BinarySearchTree::writePostorder(outFile, ptr->left);
    BinarySearchTree::writePostorder(outFile, ptr->right);
    outFile << to_string(ptr->key) << " ";
  }
}

// O(logN)
int BinarySearchTree::_getHeight(TreeNode* curr) {
  if (curr == nullptr) 
    return 0;
  return 1 + max(_getHeight(curr->left), _getHeight(curr->right));
}

void BinarySearchTree::_printSpaces(double n, TreeNode* curr) {
  for(double i = 0; i < n; ++i) 
    cout<<"   ";
  if (curr == nullptr)
    cout<<"  ";
  else
    cout<<curr->key;
}

void BinarySearchTree::printTree() {
  if (root == nullptr)
    return;
  queue<TreeNode*> q;
  q.push(root);
  queue<TreeNode*> temp;
  int cnt = 0;
  int height = _getHeight(root) - 1;
  double nMaxNodes = pow(2, height + 1) - 1;
  while (cnt <= height) {
    TreeNode* curr = q.front();
    q.pop();
    if (temp.empty())
      _printSpaces(nMaxNodes / pow(2, cnt + 1) + height - cnt, curr);
    else
      _printSpaces(nMaxNodes / pow(2, cnt), curr);
    if (curr == nullptr) {
      temp.push(nullptr);
      temp.push(nullptr);
    }
    else {
      temp.push(curr->left);
      temp.push(curr->right);
    }
    if (q.empty()) {
      cout<<endl<<endl;
      q = temp;
      queue<TreeNode*> empty;
      swap(temp, empty);
      ++cnt;
    }
  }
}

int main() {
  BinarySearchTree tree;
  ifstream inFile("input.txt");
  ofstream outFile("output.txt");
  string line;
  while (getline(inFile, line))
  {
    char op = line[0];
    istringstream iss(line.substr(1));
    int k;
    TreeNode* found = nullptr;
    vector<int> data;
    switch(op)
    {
      case BUILD:
        while (iss >> k)
          data.push_back(k);
        tree.root = tree.arrayToBST(data);
        if (tree.root) {
          outFile<<BUILD<<endl;
        }
        else {
          cerr<<"BUILD: invalid input"<<endl;
        }
        tree.printTree();
        break;

      case FIND_MIN:
        found = tree.findMin();
        if (found == nullptr) {
          cerr<<"FindMin failed"<<endl;
        }
        else {
          outFile<<found->key<<endl;
        }
        break;

      case FIND_MAX:
        found = tree.findMax();
        if (found == nullptr) {
          cerr<<"FindMax failed"<<endl;
        }
        else {
          outFile<<found->key<<endl;
        }
        break;

      case SEARCH:
        if (!(iss >> k)) {
          cerr<<"SEARCH: invalid input"<<endl;
        }
        // Practice 6. Call the function for search
        if (tree.root != nullptr) {
          if (tree.search(tree.root, k) != nullptr)
            outFile << "S " << k << endl;
        }
        break;

      case INORDER:
        // Practice 6. Call the function for inorder traversal;
        tree.writeInorder(outFile, tree.root);
        outFile << endl;
        break;

      case PREORDER:
        // Practice 6. Call the function for preorder traversal;
        tree.writePreorder(outFile, tree.root);
        outFile << endl;
        break;

      case POSTORDER:
        // Practice 6. Call the function for postorder traversal;
        tree.writePostorder(outFile, tree.root);
        outFile << endl;
        break;

      case INSERT:
        if (!(iss >> k)) {
          cerr<<"INSERT: invalid input"<<endl;
        }
        // TODO. Practice 7. Call the function for insertion
        tree.insertNode(k);
        outFile << "I " << k << endl;
        break;

      case DELETE:
        if (!(iss >> k)) {
          cerr<<"DELETE: invalid input"<<endl;
        }
        // TODO. Practice 7. Call the function for deletion
        tree.root = tree.deleteNode(tree.root, k);
        if (tree.search(tree.root, k)->key != k)
          outFile << "D " << k << endl;
        break;

      default:
        cerr<<"Undefined operator"<<endl;
    }
  }
  outFile.close();
  inFile.close();
}

File: Tree Advanced/Tree.py
# Practices 6&7. Binary Search Tree Operations
from collections import deque
BUILD = 'B'
FIND_MIN = 'm'
FIND_MAX = 'M'
SEARCH = 'S'
INSERT = 'I'
DELETE = 'D'
INORDER = 'N'
PREORDER = 'R'
POSTORDER = 'O'

# Node implementation
class TreeNode:
  def __init__(self, k, l=None, r=None):
    self.key = k
    self.left = l
    self.right = r

class BinarySearchTree:
  def __init__(self):
    self.root = None
    self.cur = self.root

  # Return True if tree is empty; False otherwise
  # Just checking, so O(1)
  def isEmpty(self):
    return self.root == None

  # Given a sequence arr of integers, build a binary search (sub)tree.
  # Return the root node of the tree
  # It is almost similar to divide part in merge sort.
  # So time complexity is O(NlogN)
  def arrayToBST(self, arr: list[int]) -> TreeNode:
    if arr != sorted(arr) or not len(arr):
      return None
    
    middle = round(len(arr)/2 - 1)
    root = TreeNode(arr[middle])
    if len(arr) == 2:
      root.right = TreeNode(arr[1])
    else:
      root.left = self.arrayToBST(arr[:middle])
      root.right = self.arrayToBST(arr[middle+1:])

    return root

  # Return the node with the minimum value 
  # in balanced tree, height is O(logN)
  # So, time complexity is O(logN)
  def findMin(self, st=None):
    if not st:
      ptr = self.root
    else:
      ptr = st
    while ptr.left != None:
      ptr = ptr.left
    return ptr

  # Return the node with the maximum value 
  # in balanced tree, height is O(logN)
  # So, time complexity is O(logN)
  def findMax(self, st=None):
    if not st:
      ptr = self.root
    else:
      ptr = st
    while ptr.right != None:
      ptr = ptr.right
    return ptr

  # O(logN)
  def _getHeight(self, curr):
    if not curr:
      return 0
    return 1 + max(self._getHeight(curr.left), self._getHeight(curr.right))

  def _printSpaces(self, n, curr):
    for _ in range(int(n)):
      print(" ", end="")
    if not curr:
      print(" ", end="")
    else:
      print(str(curr.key), end="")

  def printTree(self):
    if not self.root:
      return 
    q = deque()
    q.append(self.root)
    temp = deque()
    cnt = 0
    height = self._getHeight(self.root) - 1
    nMaxNodes = 2**(height + 1) - 1
    while cnt <= height:
      curr = q.popleft()
      if len(temp) == 0:
        self._printSpaces(nMaxNodes / (2**(cnt+1)), curr)
      else:
        self._printSpaces(nMaxNodes / (2**cnt), curr)
      if not curr:
        temp.append(None)
        temp.append(None)
      else:
        temp.append(curr.left)
        temp.append(curr.right)
      if len(q) == 0:
        print("\n")
        q = temp
        temp = deque()
        cnt += 1

  # Given a query, search for the node whose key is equal to query.
  # If the node exists, return the key
  # Otherwise, return nullptr  
  # On average, search function takes time complexity of O(logN).
  # Here, N is the number of nodes in tree.
  def search(self, st, query):
    def s(p, k):
      if not p:
        return
      if k == p.key:
        return p
      elif k < p.key and p.left != None:
        return s(p.left, k)
      elif k > p.key and p.right != None:
        return s(p.right, k)
      return p
    
    return s(st, query)

  # Given an output file, write the keys of all the nodes 
  # visited in inorder traversal
  # We need to visit all N nodes, so time complexity is O(N)
  def writeInorder(self, p, file):
    # Practice 6
    if p:
      self.writeInorder(p.left, file)
      file.write(f'{p.key} ')
      self.writeInorder(p.right, file)

  # Given an output file, write the keys of all the nodes 
  # visited in preorder traversal
  # We need to visit all N nodes, so time complexity is O(N)
  def writePreorder(self, p, file):
    # Practice 6
    # TODO
    if p:
      file.write(f'{p.key} ')
      self.writePreorder(p.left, file)
      self.writePreorder(p.right, file)
  
  # Given an output file, write the keys of all the nodes 
  # visited in postorder traversal
  # We need to visit all N nodes, so time complexity is O(N)
  def writePostorder(self, p, file):
    # Practice 6
    # TODO
    if p:
      self.writePostorder(p.left, file)
      self.writePostorder(p.right, file)
      file.write(f'{p.key} ')
  
  # If node with key k alreay exists in the tree, do nothing
  # Otherwise, insert new node with key k 
  # Find location to insert : same with search, so O(logN)
  # Adding Nodes : O(1)
  # So, time complexity is O(logN)
  def insertNode(self, p, k):
    # Practice 7
    if p == None:
      p = TreeNode(k)
    elif k < p.key:
      p.left = self.insertNode(p.left, k)
    elif k > p.key:
      p.right = self.insertNode(p.right, k)
    return p
      
  # If deletion fails, immediately terminate the program
  # Otherwise, delete the node with key k
  # It takes tree's root and key to remove, and returns root node of that subtree.
  # When calling, we just need to call like this:
  # tree.root = tree.deleteNode(tree.root, key)
  # On average, time complexity is O(logN).
  # Here, N is the number of nodes in tree.
  def deleteNode(self, r: TreeNode, k: int) -> TreeNode:
    if r == None:
      return None
    if k < r.key:
      r.left = self.deleteNode(r.left, k)
    elif k > r.key:
      r.right = self.deleteNode(r.right, k)
    else:
      if r.left == None:
        return r.right
      elif r.right == None:
        return r.left
      p = r
      while p.left != None:
        p = p.left
      r.key = p.key
      r.right = self.deleteNode(r.right, k)
    return r

if __name__ == "__main__":
  tree = BinarySearchTree()
  with open('input.txt', 'r') as inFile:
    lines = inFile.readlines()
  with open('output.txt', 'w') as outFile:
    for line in lines:
      words = line.split()
      op = words[0]
      if op == BUILD:
        data = [int(s) for s in words[1:]]
        tree.root = tree.arrayToBST(data)
        if tree.root:
          outFile.write(BUILD + "\n")
          tree.printTree()
        else:
          raise Exception("BUILD: invalid input")
        
      elif op == FIND_MIN:
        found = tree.findMin()
        if not found:
          raise Exception("FindMin failed")
        else:
          outFile.write(str(found.key) + "\n")

      elif op == FIND_MAX:
        found = tree.findMax()
        if not found:
          raise Exception("FindMax failed")
        else:
          outFile.write(str(found.key) + "\n")

      elif op == SEARCH:
        if len(words) != 2:
          raise Exception("SEARCH: invalid input")
        k = int(words[1])
        # Practice 5. Call the function for search
        found = tree.search(tree.root, k)
        if found:
          outFile.write(str(found.key) + "\n")
        else:
          raise Exception("SEARCH: Not found")
        pass

      elif op == INORDER:
        # Practice 5. Call the function for inorder traversal
        tree.writeInorder(tree.root, outFile)
        outFile.write('\n')

      elif op == PREORDER:
        # Practice 5. Call the function for preorder traversal
        tree.writePreorder(tree.root, outFile)
        outFile.write('\n')

      elif op == POSTORDER:
        # Practice 5. Call the function for postorder traversal
        tree.writePostorder(tree.root, outFile)
        outFile.write('\n')

      elif op == INSERT:
        if len(words) != 2:
          raise Exception("INSERT: invalid input")
        k = int(words[1])
        # TODO. Practice 6. Call the function for insertion
        tree.root = tree.insertNode(tree.root, k)
        if tree.search(tree.root, k):
          outFile.write(f"I {k}\n")

      elif op == DELETE:
        if len(words) != 2:
          raise Exception("DELETE: invalid input")
        k = int(words[1])
        # TODO. Practice 6. Call the function for deletion
        tree.root = tree.deleteNode(tree.root, k)
        if tree.search(tree.root, k).key != k:
          outFile.write(f"D {k}\n")

      else:
        raise Exception("Undefined operator")
        
